File: src/test/java/org/trostheide/generativeArt/core/ParameterDefinitionTest.java
package org.trostheide.generativeArt.core;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class ParameterDefinitionTest {

    @Test
    void integerFactoryCreatesCorrectDefinition() {
        ParameterDefinition pd = ParameterDefinition.integer("count", 5, 1, 10, "Number of items");

        assertEquals("count", pd.name());
        assertEquals(ParameterType.INTEGER, pd.type());
        assertEquals(5, pd.defaultValue());
        assertEquals(1, pd.min());
        assertEquals(10, pd.max());
        assertNull(pd.options());
        assertEquals("Number of items", pd.description());
    }

    @Test
    void doubleValFactoryCreatesCorrectDefinition() {
        ParameterDefinition pd = ParameterDefinition.doubleVal("scale", 1.5, 0.1, 10.0, "Scale factor");

        assertEquals("scale", pd.name());
        assertEquals(ParameterType.DOUBLE, pd.type());
        assertEquals(1.5, pd.defaultValue());
        assertEquals(0.1, pd.min());
        assertEquals(10.0, pd.max());
        assertNull(pd.options());
    }

    @Test
    void boolFactoryCreatesCorrectDefinition() {
        ParameterDefinition pd = ParameterDefinition.bool("enabled", true, "Toggle feature");

        assertEquals("enabled", pd.name());
        assertEquals(ParameterType.BOOLEAN, pd.type());
        assertEquals(true, pd.defaultValue());
        assertNull(pd.min());
        assertNull(pd.max());
        assertNull(pd.options());
    }

    @Test
    void stringFactoryCreatesCorrectDefinition() {
        ParameterDefinition pd = ParameterDefinition.string("title", "Hello", "A title");

        assertEquals("title", pd.name());
        assertEquals(ParameterType.STRING, pd.type());
        assertEquals("Hello", pd.defaultValue());
        assertNull(pd.min());
        assertNull(pd.max());
    }

    @Test
    void selectionFactoryCreatesCorrectDefinition() {
        List<String> options = List.of("A", "B", "C");
        ParameterDefinition pd = ParameterDefinition.selection("mode", "A", options, "Select mode");

        assertEquals("mode", pd.name());
        assertEquals(ParameterType.STRING, pd.type());
        assertEquals("A", pd.defaultValue());
        assertEquals(options, pd.options());
    }

    @Test
    void selectionDefaultIsInOptions() {
        List<String> options = List.of("X", "Y", "Z");
        ParameterDefinition pd = ParameterDefinition.selection("choice", "Y", options, "Pick one");

        assertTrue(pd.options().contains(pd.defaultValue().toString()));
    }
}

File: web/src/core/ParameterDefinition.test.js
import { describe, it, expect } from 'vitest';
import { ParameterDefinition } from './ParameterDefinition.js';

describe('ParameterDefinition', () => {
    it('integer factory creates correct definition', () => {
        const pd = ParameterDefinition.integer('count', 5, 1, 10, 'Number of items');

        expect(pd.name).toBe('count');
        expect(pd.type).toBe('integer');
        expect(pd.defaultValue).toBe(5);
        expect(pd.min).toBe(1);
        expect(pd.max).toBe(10);
        expect(pd.description).toBe('Number of items');
    });

    it('doubleVal factory creates correct definition', () => {
        const pd = ParameterDefinition.doubleVal('scale', 1.5, 0.1, 10.0, 'Scale factor');

        expect(pd.name).toBe('scale');
        expect(pd.type).toBe('double');
        expect(pd.defaultValue).toBe(1.5);
        expect(pd.min).toBe(0.1);
        expect(pd.max).toBe(10.0);
    });

    it('bool factory creates correct definition', () => {
        const pd = ParameterDefinition.bool('enabled', true, 'Toggle feature');

        expect(pd.name).toBe('enabled');
        expect(pd.type).toBe('boolean');
        expect(pd.defaultValue).toBe(true);
        expect(pd.min).toBeNull();
        expect(pd.max).toBeNull();
    });

    it('string factory creates correct definition', () => {
        const pd = ParameterDefinition.string('title', 'Hello', 'A title');

        expect(pd.name).toBe('title');
        expect(pd.type).toBe('string');
        expect(pd.defaultValue).toBe('Hello');
    });

    it('selection factory creates correct definition', () => {
        const pd = ParameterDefinition.selection('mode', 'A', ['A', 'B', 'C'], 'Select mode');

        expect(pd.name).toBe('mode');
        expect(pd.type).toBe('selection');
        expect(pd.defaultValue).toBe('A');
        expect(pd.options).toEqual(['A', 'B', 'C']);
    });

    it('selection default is in options', () => {
        const pd = ParameterDefinition.selection('choice', 'Y', ['X', 'Y', 'Z'], 'Pick one');
        expect(pd.options).toContain(pd.defaultValue);
    });

    it('options default to empty array', () => {
        const pd = ParameterDefinition.integer('x', 0, 0, 10, 'test');
        expect(pd.options).toEqual([]);
    });
});
